os/drivers/input/ist415.c

```diff
@@ -276,48 +428,1637 @@ static void get_touch_data(struct ist415_dev_s *priv)
+
+	ret = read(fd, firmware->data, firmware->size);
+	if (ret != firmware->size) {
+		dbg("Failed to read file: %d\n", get_errno());
+		close(fd);
+		return ERROR;
+	}
+	close(fd);
+	return OK;
+}
+
+static void release_firmware(struct firmware *firmware)
+{
+	firmware->size = 0;
+	kmm_free(firmware->data);
+
+	dbg("firmware released \n");
+}
+
+/****************************************************************************
+ * Name: ist415_isp_enable
+ ****************************************************************************/
+
+static int ist415_isp_enable(struct ist415_dev_s *dev, bool enable)
+{
+	int ret = 0;
+
+	if (enable) {
+		ret = ist415_i2c_write_isp(dev, ISP_TARGET_ENTER, ISP_CMD_ENTER, 0, NULL, 0);
+	} else {
+		ret = ist415_i2c_write_isp(dev, ISP_TARGET_EXIT, ISP_CMD_EXIT, 0, NULL, 0);
+	}
+
+	if (ret < 0) {
+		touchdbg("Fail to isp %s\n", enable ? "enable" : "disable");
+		return ret;
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: ist415_isp_erase
+ ****************************************************************************/
+
+static int ist415_isp_erase(struct ist415_dev_s *dev, uint8_t target, uint8_t area, uint32_t addr, int size)
+{
+	int count = 0;
+	int unit = 0;
+	int ret = 0;
+	int i;
+
+	touchvdbg("%s\n", __func__);
+
+	if (area == ISP_CMD_MASS_ERASE) {
+		count = 1;
+	} else if (area == ISP_CMD_SECTOR_ERASE) {
+		count = size / IST415_ROM_SECTOR_SIZE;
+		unit = IST415_ROM_SECTOR_SIZE;
+	} else if (area == ISP_CMD_PAGE_ERASE) {
+		count = size / IST415_ROM_PAGE_SIZE;
+		unit = IST415_ROM_PAGE_SIZE;
+	}
+
+	for (i = 0; i < count; i++) {
+		ret = ist415_i2c_write_isp(dev, target, area, addr + (i * unit), NULL, 0);
+		if (ret < 0) {
+			touchdbg("Fail to erase ISP\n");
+		}
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: ist415_isp_program
+ ****************************************************************************/
+
+static int ist415_isp_program(struct ist415_dev_s *dev, uint8_t target, uint32_t addr, uint32_t *buf32, int size)
+{
+	int ret = 0;
+	int i;
+
+	touchvdbg("%s\n", __func__);
+
+	for (i = 0; i < (size / IST415_ROM_SECTOR_SIZE); i++) {
+		ret = ist415_i2c_write_isp(dev, target, ISP_CMD_PROG, addr, buf32, IST415_ROM_SECTOR_SIZE / sizeof(uint32_t));
+		if (ret < 0) {
+			touchdbg("Fail to write ISP (Sector %d)\n", i);
+			break;
+		}
+		addr += IST415_ROM_SECTOR_SIZE;
+		buf32 += (IST415_ROM_SECTOR_SIZE / sizeof(uint32_t));
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: ist415_isp_fw_update
+ ****************************************************************************/
+
+static int ist415_isp_fw_update(struct ist415_dev_s *dev, const uint8_t *buf)
+{
+	FAR struct touchscreen_s *upper = dev->upper;
+	int ret = 0;
+
+	touchvdbg("%s\n", __func__);
+
+	ist415_reset(upper, true);
+
+	ret = ist415_isp_enable(dev, true);
```

Comment:
I think this isp should be disabled also when exit or in return
ist415_isp_enable(dev, false);